kube/src/client/discovery.rs

```diff
@@ -0,0 +1,341 @@
+            .into_iter()
+            .find(|gvk| gvk.kind == kind)
+            .map(|gvk| {
+                let data = group
+                    .versions_and_resources
+                    .iter()
+                    .find(|data| data.version == version)
+                    .unwrap();
+                let raw = data.list.resources.iter().find(|r| r.kind == kind).unwrap();
+                (gvk, raw.clone())
+            })
+    }
+}
+
+impl Group {
+    /// Core group name
+    pub const CORE_GROUP: &'static str = "core";
+
+    /// Returns the name of this group.
+    /// For core group (served at `/api`), returns "core" (also declared as
+    /// `Group::CORE`).
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Actually sets up order promised by `version`
+    fn sort_versions(&mut self) {
+        self.versions_and_resources
+            .sort_by_cached_key(|ver_data| Version::parse(ver_data.version.as_str()))
+    }
+
+    /// Returns served versions (e.g. `["v1", "v2beta1"]`) of this group.
+    /// This list is always non-empty, and sorted in the following order:
+    /// - Stable versions (with the last being the first)
+    /// - Beta versions (with the last being the first)
+    /// - Alpha versions (with the last being the first)
+    /// - Other versions, alphabetically
+    // Order is documented here:
+    // https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions
+    pub fn versions(&self) -> impl Iterator<Item = &str> {
+        let versions = self.versions_and_resources.as_slice();
+        versions.iter().map(|ver_data| ver_data.version.as_str())
+    }
+
+    /// Returns preferred version for working with given group.
+    /// This function tries to use `preferredVersion` field at first,
+    /// and then uses some fallbacks if that is missing.
+    pub fn preferred_version(&self) -> &str {
+        match &self.preferred_version {
+            Some(v) => v,
+            None => self.versions().next().unwrap(),
+        }
+    }
+
+    /// Returns resources available in version `ver` of this group.
+    /// If the group does not support this version,
+    /// returns empty vector.
+    pub fn resources_by_version(&self, ver: &str) -> Vec<GroupVersionKind> {
+        let resources = self
+            .versions_and_resources
+            .iter()
+            .find(|ver_data| ver_data.version == ver)
+            .map(|ver_data| ver_data.resources.as_slice())
+            .unwrap_or(&[]);
+        resources
+            .iter()
+            .cloned()
+            .map(|mut api_resource| {
+                api_resource.group = Some(if self.name == "core" {
+                    String::new()
+                } else {
+                    self.name.clone()
+                });
+                api_resource.version = Some(ver.to_string());
+                GroupVersionKind::from_api_resource(&api_resource, "unused/v0")
```

Comment:
why can we get away with passing `unused/v0` here?

Reply:
We explicitly set `group` and `version`, so `from_api_resource` will ignore the second argument.
I'll add a comment clarifying this.